File: elixir/test/rpn_calc/calculator_test.exs
defmodule RPNCalculator.CalculatorTest do
  use ExUnit.Case

  alias RPNCalc.Calculator

  setup do
    Agent.update(Calculator, fn _ -> [] end)
  end

  describe "process/1 with invalid inputs" do
    test "raises error on invalid input" do
      assert_raise RuntimeError, "Invalid input.", fn ->
        Calculator.process("%")
      end
    end

    test "raises error on operations with too few values" do
      Calculator.process("2")

      assert_raise RuntimeError, "At least 2 values are required.", fn ->
        Calculator.process("-")
      end
    end

    test "raises error on division by 0" do
      Calculator.process("5")
      Calculator.process("0")

      assert_raise RuntimeError, "Function returns Infinity.", fn ->
        Calculator.process("/")
      end
    end
  end

  describe "process/1" do
    # standard arithmetic operators
    test "supports addition" do
      asserts = ~w(5 8 13)

      ~w(5 8 +)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end

    test "supports subtraction" do
      asserts = ~w(5 8 -3)

      ~w(5 8 -)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end

    test "supports multiplication" do
      asserts = ~w(5 8 40)

      ~w(5 8 *)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end

    test "supports division" do
      asserts = ~w(5 8 0.625)

      ~w(5 8 /)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end

    # negative and decimal numbers

    test "supports negative numbers" do
      asserts = ~w(-5 -8 -13)

      ~w(-5 -8 +)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end

    test "supports decimal numbers" do
      asserts = ~w(-0.5 -8 4.0)

      ~w(-0.5 -8 *)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end

    # number of operations

    test "supports stacked operators" do
      asserts = ~w(2 9 3 12 24)

      ~w(2 9 3 + *)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end

    test "supports chained operations" do
      asserts = ~w(20 13 7 2 3.5)

      ~w(20 13 - 2 /)
      |> Enum.with_index()
      |> Enum.each(fn {x, i} ->
        assert Calculator.process(x) == Enum.at(asserts, i)
      end)
    end
  end
end

File: go/internal/pkg/calclib/calculator_test.go
// Package calclib implements an RPN calculator.
package calclib

import (
	"errors"
	"reflect"
	"testing"
)

func TestProcess(t *testing.T) {
	type args struct {
		input string
	}
	tests := []struct {
		name       string
		args       args
		wantResult string
		wantErr    error
	}{
		{
			"returns warning on invalid input",
			args{"%"},
			"",
			errors.New("invalid input"),
		},
		{
			"returns warning on operations with too few values",
			args{"+"},
			"",
			errors.New("at least 2 values are required"),
		},
		{
			"returns value on valid input",
			args{"1"},
			"1",
			nil,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			gotResult, err := Process(tt.args.input)
			if (err != nil) && !reflect.DeepEqual(err, tt.wantErr) {
				t.Errorf("Process() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if gotResult != tt.wantResult {
				t.Errorf("Process() = %v, want %v", gotResult, tt.wantResult)
			}
		})
	}

	t.Run("returns warning on division by 0", func(t *testing.T) {
		_, _ = Process("5")
		_, _ = Process("0")
		gotResult, err := Process("/")

		wantResult := ""
		wantErr := errors.New("function returns infinity")
		if (err != nil) && !reflect.DeepEqual(err, wantErr) {
			t.Errorf("Process() error = %v, wantErr %v", err, wantErr)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports addition", func(t *testing.T) {
		_, _ = Process("5")
		_, _ = Process("8")
		gotResult, err := Process("+")
		wantResult := "13"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports subtraction", func(t *testing.T) {
		_, _ = Process("5")
		_, _ = Process("8")
		gotResult, err := Process("-")
		wantResult := "-3"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports multiplication", func(t *testing.T) {
		_, _ = Process("5")
		_, _ = Process("8")
		gotResult, err := Process("*")
		wantResult := "40"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports division", func(t *testing.T) {
		_, _ = Process("5")
		_, _ = Process("8")
		gotResult, err := Process("/")
		wantResult := "0.625"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports negative numbers", func(t *testing.T) {
		_, _ = Process("-5")
		_, _ = Process("-8")
		gotResult, err := Process("+")
		wantResult := "-13"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports decimal numbers", func(t *testing.T) {
		_, _ = Process("-0.5")
		_, _ = Process("-8")
		gotResult, err := Process("*")
		wantResult := "4"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports stacked operators", func(t *testing.T) {
		_, _ = Process("2")
		_, _ = Process("9")
		_, _ = Process("3")
		_, _ = Process("+")
		gotResult, err := Process("*")
		wantResult := "24"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})

	t.Run("supports chained operations", func(t *testing.T) {
		_, _ = Process("20")
		_, _ = Process("13")
		_, _ = Process("-")
		_, _ = Process("2")
		gotResult, err := Process("/")
		wantResult := "3.5"
		if err != nil {
			t.Errorf("Process() error = %v, wantErr %v", err, nil)
			return
		}
		if gotResult != wantResult {
			t.Errorf("Process() = %v, want %v", gotResult, wantResult)
		}
	})
}